### ISSUE TITLE
Compatibility with devices built for <MSA-1.3

### ISSUE THREAD

**Issue by alexrkaufman** [alexrkaufman/pytla#2]
## The Problem

* There are many devices made prior to the creation of the 1.3 standard and it would be beneficial if pytla could control these devices as well. In these cases incompatible registers cause `Register Not Implemented` errors. 
* We have encountered devices that have firmware developed for <MSA-1.3 that were nevertheless built for MSA-1.3. In these cases incompatible registers do not cause `Register Not Implemented` errors however they nevertheless cause other errors.  

There are two conflicting considerations we must grapple with. The first is user experience and the second is developer experience. For user experience it is better to have similar patterns across all versions but it is also better to make this compatibility available sooner than later. For developer experience I would like to make sure that the code is as clear as possible.  To me this means avoiding copious `if ... else` statements to check control flow for various versions. 

## Possible solution 1

Create separate classes for each version of the standard with their own version of the `yaml` register files. In this case the classes available would be `ITLA` for the 1.3 compliant devices, and `ITLA10`, `ITLA11`, and `ITLA12` for MSA-1.0 through 1.2. 

The downside of this approach is in UX. There isnt a single pattern for users to follow and they will likely use the ITLA class by default and run into errors. We will have to update the readme to mention that they should change the initialization process to use the alternate classes. 

Upside is that it avoids if statements to 

## Possible solution 2

Add a parameter to `__init__`  in the `ITLA` class allowing the user to select which version of the MSA they would like to use. This will require a lot of `if...else` statements in functions like `get_frequency` to avoid registers that are not compatible from version to version. This makes for a more difficult developer experience and a simplified user experience. 

## Possible solution 3

disclaimer: I don't know how one would implement this. 

This is basically a best of both worlds scenario. If it is possible then this would be my preferred option.

The `ITLA` class returns other classes based on user suggestions, defaulting to 1.3. In this scenario we would build out classes and yaml files for each MSA-1.X all named `ITLA1X`. In initialization the user could choose to provide a version number as a parameter which would then select the appropriate class. 

The `ITLA` class would then define a bunch of abstract methods that the subclasses would have to implement. This way a cohesive interface could be maintained across all `ITLA` versions. The `__init__.py` file in `pyserial` is a potential model for this approach. [here](https://github.com/pyserial/pyserial/blob/master/serial/__init__.py)

This is another possible approach I think(?) [ABC](https://docs.python.org/3/library/abc.html).

I think this provides the best of both worlds for user and developer experience by removing alot of control flow and providing a single point of entry. 

## Where to go from here? 

This is a comprehensive account of my current thoughts both pro and con for each approach. It seems to me that for the time being the thing to do is to implement "Possible Solution 1" and in a later version build the single point of entry (potentially this would be a good impetus for a 1.0 release?)

**Comment by alexrkaufman** [comment]
The develop branch as of commit aa8990d has a version of the interface defined in **Possible Solution 3**. This has been achieved by implementing an `ITLABase` class and a very sparse `ITLA` class used for selecting the appropriate subclass/version.

I have not yet tested this on communication with actual 1.3 and 1.2 based lasers. Tests so far have been on checking whether the correct version subclass is selected.

Once this has been tested on 1.2 and 1.3 lasers with version selection we will close this issue. The 1.2 and purephotonics implementations must avoid 1.3 register functions (0x63-0x6B) which I think I have done but must verify.

**Comment by cyangy** [comment]
Hello, @alexrkaufman , I have tested the develop branch and it work's like a charm.
```python3
laser = itla.ITLA('COM4', 9600, version='1.2')
```
 Except this:
```python3
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\cyangy\Python\lib\site-packages\itla\itla12.py", line 500, in get_frequency_min
    response = self._lfl3()
AttributeError: 'ITLA12' object has no attribute '_lfl3'
```

**Comment by alexrkaufman** [comment]
@cyangy, This should be fixed as of ca4703f. Glad it was mostly working! I will merge ca4703f into main in a couple of days so I have time to test myself and to give time for other bugs to pop up. Please feel free to submit issues as you run into them. 

This is unrelated to this issue but: Please note that the function `enable()` has had a slight modification in behavior. The `CPException` is no longer caught in the enable function. Users must do that themselves explicitly as shown in the updated readme. The reasoning behind this is that the lasers do not enable immediately and hiding the command pending exception could result in the script progressing past what the laser is capable of doing at that point in time. Instead users are encouraged to catch the exception themselves and add a time delay. 

I think there may be a better way to do this. Possibly by reading the pending bit in `nop`. Happy to open a new issue about it if you have an opinion or suggestions.

**Comment by alexrkaufman** [comment]
I did a little reworking/squashing of the history in develop because it became pretty messy. commit 810c075 is the version that has been merged into main tagged version 0.1.0. This has been updated and uploaded to PyPI. commit ca4703f will no longer exist though. `pip install -U pytla` should update you to the most recent version. 

Closing this issue.
